drivers/src/stm32f446xx_usart_driver_course.c

```diff
@@ -110,3 +153,63 @@ void USART_WriteData(USART_Handler_TypedDef* pUSART_Handler, uint8_t* pUSART_TX_
 	{
 	}
 }
+
+void USART_SetBaudRate(USART_Handler_TypedDef* pUSART_Handler)
+{
+	uint16_t tempreg, mantissa, DIV_fraction, DIV_mantissa;
+	float fraction;
+
+	uint32_t usart_clock, usart_div;
+
+	//UART1 and USART6 are on APB2 so they use PCLK2
+	if((pUSART_Handler->pUSART == UART1) || (pUSART_Handler->pUSART == UART6))
+	{
+		usart_clock = RCC_GetPCLK2();
+	}
+	else
+	{
+		usart_clock = RCC_GetPCLK1();
+	}
+
+	usart_div = (usart_clock*100)/((8*2-pUSART_Handler->USART_Pin_Config.USART_OverSampling)*pUSART_Handler->USART_Pin_Config.USART_BaudRate);
+
+	mantissa = usart_div/100;
+	fraction = usart_div - (mantissa*100);
+	fraction = fraction/100;
+
+	if(pUSART_Handler->USART_Pin_Config.USART_OverSampling == 0)	//oversampling = 16, fraction coded on 4 bits
+	{
+		DIV_fraction = (fraction*16) + 0.5; //round to the nearest number
```

Comment:
Here,
this calculation will take aproximately 300 clock cycles.

All implicit numbers in Float calculations are converted to Double with the functions below.
    __aeabi_ui2d: convert unsigned int to double
    __aeabi_dmul: double multiplication
    __aeabi_ddiv: double division
    __aeabi_f2d: float to double conversion
    __aeabi_dsub: double subtraction
    __aeabi_d2f: double to float conversion

So what happens here is that your PC will use 64bits, and require your ARM 32bits processor to calculate this, the only way is to use the Software FPU, and this is very problematic.

You can implicit convert the literals like:
DIV_fraction = (fraction*8f) + 0.5f;	

Or include the below flag in the compiler
-fsingle-precision-constant